lib/db_connection/ownership/manager.ex

```diff
@@ -77,11 +80,25 @@ defmodule DBConnection.Ownership.Manager do
     {:ok, pool, owner_sup} = PoolSupervisor.start_pool(module, pool_opts)
     mode = Keyword.get(pool_opts, :ownership_mode, :auto)
     {:ok, %{pool: pool, owner_sup: owner_sup, checkouts: %{}, owners: %{},
-            mode: mode, ets: ets}}
+            mode: mode, mode_ref: nil, ets: ets}}
   end
 
-  def handle_call({:mode, mode}, _from, state) do
-    {:reply, :ok, %{state | mode: mode}}
+  def handle_call({:mode, {:shared, pid}}, _from, %{mode: {:shared, current}} = state) do
+    cond do
+      pid == current ->
+        {:reply, :ok, state}
+      Process.alive?(current) ->
+        {:reply, :already_shared, state}
+      true ->
+        share_and_reply(state, pid)
+    end
+  end
+  def handle_call({:mode, {:shared, pid}}, _from, state) do
+    share_and_reply(state, pid)
+  end
+  def handle_call({:mode, mode}, _from, %{mode_ref: ref} = state) do
+    ref && Process.demonitor(ref, [:flush])
```

Comment:
I don't think we want to demonitor here because owner must still be in checkouts.